Class/CDC/ECM/usbh_ecm.c

```diff
@@ -328,20 +330,105 @@ USBH_ERR  USBH_CDC_ECM_Remove (USBH_CDC_ECM_DEV  *p_cdc_ecm_dev)
         return USBH_ERR_INVALID_ARG;
     }
 
+    p_cdc_dev = p_cdc_ecm_dev->CDC_DevPtr;
+
     p_cdc_ecm_dev->EventNotifyPtr    = p_ecm_event_notify;
     p_cdc_ecm_dev->EventNotifyArgPtr = p_arg;
 
-    return USBH_CDC_EventNotifyReg(        p_cdc_ecm_dev->CDC_DevPtr,
-                                           USBH_CDC_ECM_EventRxCmpl,
+    err = USBH_IntrRxAsync(       &p_cdc_dev->CIC_IntrIn,
+                           (void *)p_cdc_dev->EventNotifyBuf,
+                                   USBH_CDC_LEN_EVENT_BUF,
+                                   USBH_CDC_ECM_EventRxCmpl,
                                    (void *)p_cdc_ecm_dev);
+    if (err != USBH_ERR_NONE) {
+        (void)USBH_EP_Reset(p_cdc_dev->DevPtr,
+                           &p_cdc_dev->CIC_IntrIn);
+
+        if (err == USBH_ERR_EP_STALL) {
+            (void)USBH_EP_StallClr(&p_cdc_dev->CIC_IntrIn);
+        }
+    }
+
+    return (err);
+}
+
+
+/*
+*********************************************************************************************************
+*                                          USBH_CDC_ECM_DataTx()
+*
+* Description : Send ECM data to device.
+*
+* Argument(s) : p_cdc_ecm_dev     Pointer to CDC ECM device.
+*
+*               p_buf             Pointer to buffer of data that will be sent.
+*
+*               buf_len           Buffer length in octets.
+*
+*               timeout_ms        Timeout, in milliseconds.
+*
+*               p_err             Variable that will receive the return error code from this function.
+*
+*                                                                   ----- RETURNED BY USBH_BulkTx() : -----
+*                           USBH_ERR_NONE                           Bulk transfer successfully transmitted.
+*                           USBH_ERR_INVALID_ARG                    Invalid argument passed to 'p_ep'.
+*                           USBH_ERR_EP_INVALID_TYPE                Endpoint type is not Bulk or direction is not OUT.
+*                           USBH_ERR_EP_INVALID_STATE               Endpoint is not opened.
+*                           USBH_ERR_NONE,                          URB is successfully submitted to host controller.
+*                           Host controller drivers error code,     Otherwise.
+*
+* Return(s)   : Number of octets transferred.
+*
+* Note(s)     : None.
+*********************************************************************************************************
+*/
+
+CPU_INT32U  USBH_CDC_ECM_DataTx (USBH_CDC_ECM_DEV  *p_cdc_ecm_dev,
+                                 CPU_INT08U        *p_buf,
+                                 CPU_INT32U         buf_len,
+                                 CPU_INT32U         timeout_ms,
+                                 USBH_ERR          *p_err)
+{
+    CPU_INT32U  xfer_len;
+    USBH_CDC_DEV  *p_cdc_dev;
+
+
+    if (p_cdc_ecm_dev == (USBH_CDC_ECM_DEV *)0) {
+       *p_err = USBH_ERR_INVALID_ARG;
+        return (0u);
+    }
+
```

Comment:
Add a check for `p_buf == CPU_INT08U *)0` and `buf_len == 0`

```
    if (buf_len == 0u) {
       *p_err = USBH_ERR_INVALID_ARG;
        return (0u);
    }

    if (p_buf == (CPU_INT08U *)0) {
       *p_err = USBH_ERR_NULL_PTR;
        return (0u);
    }
```

Reply:
I can change it, but I wanted to stay at parity with what `USBH_CDC_DataTx` is doing and it seems to allow sending 0-length data.

Reply:
I added it, but curious if that could cause some confusion with the `USBH_CDC_DataTx` function.

Reply:
I see what you mean. I was trying to follow the `USBH_FTDI_Tx()` function, because I was not sure if a 0 length data is allowed for ECM. Let's remove it then, I don't think there will be any harm

Reply:
SGTM!